Ordinal numbers scatter 15

francares opened the pull request.

mcovarr commented on src/main/scala/cromwell/backend/model/TaskDescriptor.scala:
```diff
@@ -15,6 +16,7 @@ import wdl4s.values.WdlValue
   * @param runtimeAttributes Configuration needed at runtime in order to execute a task.
   */
 case class TaskDescriptor(name: String,
+                          index: Option[Int],
```
There's an `ExecutionIndex` type for this in Cromwell, that probably should be brought over here.

francares replied:
Done.

mcovarr commented on src/main/scala/cromwell/backend/provider/local/LocalBackend.scala:
```diff
@@ -314,22 +317,25 @@ class LocalBackend(task: TaskDescriptor) extends BackendActor with StrictLogging
     val stderrFileLength = Try(Files.size(stderr)).getOrElse(0L)
     val failOnErrFlag = failOnStderr.getOrElse("false").toBoolean
-    val rc = returnCode.contentAsString.stripLineEnd.toInt
+    val rc = Try(returnCode.contentAsString.stripLineEnd.toInt)
```
Not a big deal on the local backend, but this could be made lazy so none of this work is done if the `processReturnCode` is nonzero.  This is a bigger deal on non-local backends where that file would need to be downloaded but here it's just a nicety.  :smile:

francares replied:
Fixed.

mcovarr commented on src/main/scala/cromwell/backend/provider/local/LocalBackend.scala:
```diff
@@ -314,22 +317,25 @@ class LocalBackend(task: TaskDescriptor) extends BackendActor with StrictLogging
+    } else if (rc.isFailure) { // case where docker fails.
+      TaskFinalStatus(Status.Failed, FailureTaskResult(
+        new IllegalStateException("Could not get RC file."), processReturnCode, stderr.toString))
+    } else if (failOnErrFlag && stderrFileLength > 0) { // rc status is validated in previous step so it is safe to use .get
```
comment makes more sense moved down a line

francares replied:
Moved down.

mcovarr commented on src/main/scala/cromwell/backend/provider/local/LocalBackend.scala:
```diff
@@ -314,22 +317,25 @@ class LocalBackend(task: TaskDescriptor) extends BackendActor with StrictLogging
+    } else if (rc.get != 0 && !isInContinueOnReturnCode(rc.get)) {
       TaskFinalStatus(Status.Failed, FailureTaskResult(
-        new IllegalStateException("Return code is not equals to zero."), rc, stderr.toString))
+        new IllegalStateException("Return code is not equals to zero."), rc.get, stderr.toString))
```
not equals to zero -> nonzero

mcovarr replied:
also wouldn't hurt to include the actual return code in the message

francares replied:
Done.

mcovarr commented on src/main/scala/cromwell/backend/provider/local/LocalBackend.scala:
```diff
@@ -314,22 +317,25 @@ class LocalBackend(task: TaskDescriptor) extends BackendActor with StrictLogging
-    } else if (failOnErrFlag && stderrFileLength > 0) {
+    } else if (rc.isFailure) { // case where docker fails.
+      TaskFinalStatus(Status.Failed, FailureTaskResult(
+        new IllegalStateException("Could not get RC file."), processReturnCode, stderr.toString))
```
There's possibly a message in the exception associated with the failure that isn't being logged or included in the generated `IllegalStateException`.

mcovarr commented on src/main/scala/cromwell/backend/provider/local/LocalBackend.scala:
```diff
@@ -314,22 +317,27 @@ class LocalBackend(task: TaskDescriptor) extends BackendActor with StrictLogging
+    } else if (rc.isFailure) {
+      // case where docker fails.
+      TaskFinalStatus(Status.Failed, FailureTaskResult(
+        new IllegalStateException("Could not get RC file."), processReturnCode, stderr.toString))
```
I commented on an older version of this and it disappeared from view, apologies if you see this twice. :smile: 

The exception associated with the `rc.isFailure` case should be given as the cause for this `IllegalStateException`.

francares replied:
Done.

mcovarr commented:
:+1: minus the one exception nit, the wheel chooses @cjllanwarne

mcovarr commented:
Will you also update Cromwell on pluggable_backends_develop to reference this updated cromwell-backend artifact?

mcovarr commented:
Please squash and rebase this before merge.

cjllanwarne commented:
:+1: 

Nothing to do with these changes, just opining into the wind - the conflating of "processReturnCode" and "rc" makes me uncomfortable. For the same reason that mixing JES error codes into "rc" ended up making things awkward.

francares committed:
Added ordinal index for scatter task executions.

Restore ordinal scatter

Added ordinal index for scatter task executions.

Restore ordinal scatter
